File: ts/project/.sdk/tm/php/feature/StreamingFeature.php
<?php
declare(strict_types=1);

// ProjectName SDK streaming feature

require_once __DIR__ . '/BaseFeature.php';

// Streaming result support. For list-style operations it attaches a
// `$result->stream` closure returning a PHP Generator so callers can
// consume items incrementally with `foreach (($result->stream)() as $item)`
// instead of materialising the whole array. The generator reads the
// result's data lazily, so it reflects the parsed entities. A `chunkDelay`
// (ms) simulates paced/chunked delivery via the injectable `sleep`; a
// `chunkSize` groups items into batches when set. Mirrors
// ts/src/feature/streaming/StreamingFeature.ts.
class ProjectNameStreamingFeature extends ProjectNameBaseFeature
{
    private mixed $client;
    private ?array $options;

    public function __construct()
    {
        parent::__construct();
        $this->version = '0.0.1';
        $this->name = 'streaming';
        $this->active = true;
        $this->client = null;
        $this->options = null;
    }

    public function init(ProjectNameContext $ctx, array $options): void
    {
        $this->client = $ctx->client;
        $this->options = $options;
        $this->active = ($options['active'] ?? null) === true;
    }

    public function PreResult(ProjectNameContext $ctx): void
    {
        if (!$this->active || !$this->_streamable($ctx)) {
            return;
        }
        $result = $ctx->result;
        if ($result === null) {
            return;
        }

        $result->streaming = true;
        $result->stream = function () use ($result): \Generator {
            return $this->_iterate($result);
        };

        $client = $this->client;
        if (!isset($client->_streaming)) {
            $client->_streaming = ['opened' => 0];
        }
        $client->_streaming['opened']++;
    }

    private function _iterate(ProjectNameResult $result): \Generator
    {
        $chunk_delay = is_numeric($this->options['chunkDelay'] ?? null)
            ? (float)$this->options['chunkDelay'] : 0.0;
        $chunk_size = is_numeric($this->options['chunkSize'] ?? null)
            ? (int)$this->options['chunkSize'] : 0;

        // Read lazily so downstream result processing is reflected.
        $items = is_array($result->resdata) ? array_values($result->resdata) : [];

        if ($chunk_size > 0) {
            $count = count($items);
            for ($i = 0; $i < $count; $i += $chunk_size) {
                if ($chunk_delay > 0) {
                    $this->_sleep($chunk_delay);
                }
                yield array_slice($items, $i, $chunk_size);
            }
            return;
        }

        foreach ($items as $item) {
            if ($chunk_delay > 0) {
                $this->_sleep($chunk_delay);
            }
            yield $item;
        }
    }

    private function _streamable(ProjectNameContext $ctx): bool
    {
        $ops = $this->options['ops'] ?? ['list'];
        return is_array($ops) && in_array($ctx->op->name, $ops, true);
    }

    private function _sleep(mixed $ms): void
    {
        if (!is_numeric($ms) || $ms <= 0) {
            return;
        }
        $sleep = $this->options['sleep'] ?? null;
        if (is_callable($sleep)) {
            $sleep($ms);
            return;
        }
        usleep((int)($ms * 1000));
    }
}

File: ts/project/.sdk/tm/lua/feature/streaming_feature.lua
-- ProjectName SDK streaming feature
--
-- Streaming result support. For list-style operations it attaches a
-- `result.stream` function (PreResult) returning a stateful Lua iterator
-- so callers can consume items incrementally:
--
--   for item in result.stream() do ... end
--
-- The iterator reads the result's `resdata` lazily (at `stream()` call
-- time), so it reflects downstream result processing. A `chunkDelay` (ms)
-- simulates paced/chunked delivery for offline tests via the injectable
-- `sleep`; a `chunkSize` groups items into batch tables when set.

local BaseFeature = require("feature.base_feature")

local StreamingFeature = {}
StreamingFeature.__index = StreamingFeature
setmetatable(StreamingFeature, { __index = BaseFeature })


function StreamingFeature.new()
  local self = setmetatable(BaseFeature.new(), StreamingFeature)
  self.version = "0.0.1"
  self.name = "streaming"
  self.active = true
  self.client = nil
  self.options = nil
  return self
end


function StreamingFeature:init(ctx, options)
  self.client = ctx.client
  self.options = options or {}

  if options["active"] == true then
    self.active = true
  else
    self.active = false
  end
end


function StreamingFeature:PreResult(ctx)
  if not self.active then
    return
  end
  if not self:_streamable(ctx) then
    return
  end
  local result = ctx.result
  if result == nil then
    return
  end

  local stream_self = self
  result.streaming = true

  result.stream = function()
    return stream_self:_iterate(result)
  end

  local client = self.client
  if client._streaming == nil then
    client._streaming = { opened = 0 }
  end
  client._streaming.opened = client._streaming.opened + 1
end


-- Build a stateful closure iterating the result items (or batches of
-- `chunkSize` items), pacing each step by `chunkDelay` ms.
function StreamingFeature:_iterate(result)
  local chunk_delay = self.options["chunkDelay"] or 0
  local chunk_size = self.options["chunkSize"] or 0

  -- Read lazily so downstream result processing is reflected.
  local items = result.resdata
  if type(items) ~= "table" then
    items = {}
  end

  local i = 0

  if chunk_size > 0 then
    return function()
      if i >= #items then
        return nil
      end
      if chunk_delay > 0 then
        self:_sleep(chunk_delay)
      end
      local batch = {}
      local last = math.min(i + chunk_size, #items)
      for j = i + 1, last do
        table.insert(batch, items[j])
      end
      i = last
      return batch
    end
  end

  return function()
    if i >= #items then
      return nil
    end
    if chunk_delay > 0 then
      self:_sleep(chunk_delay)
    end
    i = i + 1
    return items[i]
  end
end


function StreamingFeature:_streamable(ctx)
  local ops = self.options["ops"] or { "list" }
  local opname = ctx.op ~= nil and ctx.op.name or nil
  for _, o in ipairs(ops) do
    if o == opname then
      return true
    end
  end
  return false
end


function StreamingFeature:_sleep(ms)
  if ms == nil or ms <= 0 then
    return
  end
  local sleep = self.options["sleep"]
  if type(sleep) == "function" then
    sleep(ms)
    return
  end
  -- Portable stdlib-only wait: spin on the CPU clock.
  local target = os.clock() + (ms / 1000)
  while os.clock() < target do end
end


return StreamingFeature
